paasta_tools/check_spark_jobs.py

```diff
@@ -0,0 +1,170 @@
+    for key, value in properties:
+        if key == 'spark.executorEnv.PAASTA_SERVICE':
+            service = value
+            break
+    else:
+        return None
+    if service.startswith(JUPYTER_PREFIX):
+        return service[len(JUPYTER_PREFIX):]
+    else:
+        return service
+
+
+def get_matching_framework_info(min_hours):
+    frameworks = mesos_tools.get_all_frameworks(active_only=True)
+    matching_info = []
+    min_timedelta = datetime.timedelta(hours=min_hours)
+    for framework in frameworks:
+        if not framework.active:
+            continue
+        if framework.get('principal') != 'spark':
+            continue
+        time_running = get_time_running(framework)
+        if time_running >= min_timedelta:
+            info = {
+                'id': framework.id,
+                'name': framework.name,
+                'webui_url': framework.get('webui_url'),
+                'service': guess_service(get_spark_properties(framework)),
+                'user': framework.user,
+                'time_running': str(time_running),
+            }
+            matching_info.append(info)
+
+    return matching_info
+
+
+def format_framework(info):
+    result = [f'{info["name"]} (running for {info["time_running"]})']
+    result.append(f'  user: {info["user"]}')
+    result.append(f'  job UI: {info["webui_url"]}')
+    return '\n'.join(result)
+
+
+def format_message_for_service(service, frameworks):
+    output = f'Found the following long-running Spark frameworks associated with service {service}.\n'
+    output += f'Please check why they are still running and terminate if appropriate.\n\n'
+    output += '\n'.join(format_framework(f) for f in frameworks)
+    return output
+
+
+def get_messages_by_service(results):
+    results_by_service = defaultdict(list)
+    for result in results:
+        service = result['service']
+        results_by_service[service].append(result)
+
+    return {
+        service: format_message_for_service(service, results)
+        for service, results in results_by_service.items()
+    }
+
+
+def notify_framework_owners(service, output):
+    overrides = {
+        'page': False,
+        'alert_after': 0,
+        'tip': 'Ask the user to check the job UI and terminate the job if appropriate.',
+        'runbook': 'http://y/spark-debug',
+    }
+    send_event(
+        service=service,
+        check_name=f'long_running_spark_jobs.{service}',
+        overrides=overrides,
+        status=1,
+        output=output,
+        soa_dir=DEFAULT_SOA_DIR,
+    )
+
+
+def report_spark_jobs(min_hours, no_notify):
+    results = get_matching_framework_info(
+        min_hours=min_hours,
+    )
+    messages_by_service = get_messages_by_service(results)
+
+    print('\n\n'.join(messages_by_service.values()))
+
+    if not no_notify:
+        valid_services = set(list_services())
+        for service, message in messages_by_service.items():
+            if service in valid_services:
```

Comment:
nit: given that `guess_service` can return `None`, maybe `get_matching_framework_info` or `get_messages_by_service` should filter those out before we get here.   Currently I think the way this'll go is that we'll get some entries in `messages_by_service` for a `None` service and then say that we didn't notify for service: `None`.

It'd maybe be nicer to handle that case explicitly: notice that we couldn't figure out the service for a framework and report that it couldn't figure out the service for the framework with name x, webui_url y.

Unsure how likely this is to happen, though.

Reply:
I think it will be relatively common at first. One case is where the service is None, If we're unable to get the value of PAASTA_SERVICE through the Spark URL. It can also be invalid if PAASTA_SERVICE has a malformed value, which is pretty common in our jupyter setups. For example, the cookiecutter jupyter configs add the user name in some cases, or use dashes instead of underscores for the service name.

I separated these out into a separate section in the report output, so it's more clear that we need to take action on them ourselves. It includes the name and url. All None messages would show up there too, under one heading for "service None". Does that address your concern?

Here's what some test output looks like:
```
Found the following long-running Spark frameworks associated with service service1.
Please check why they are still running and terminate if appropriate.

spark1 (running for 7:00:00)
  user: test_user
  job UI: url1

Found the following long-running Spark frameworks associated with service service2.
Please check why they are still running and terminate if appropriate.

spark2 (running for 7:00:00)
  user: test_user
  job UI: url2
spark3 (running for 7:00:00)
  user: test_user
  job UI: url3


INVALID SERVICES
----------------
The following frameworks are associated with services that are not configured in PaaSTA.

Found the following long-running Spark frameworks associated with service service_dne.
Please check why they are still running and terminate if appropriate.

spark3 (running for 7:00:00)
  user: test_user
  job UI: url3
```